swift/apple/FirezoneKit/Sources/FirezoneKit/Stores/Store.swift

```diff
@@ -46,89 +49,110 @@ public final class Store: ObservableObject {
       }
     }
 
+    // Load our state from the system. Based on what's loaded, we may need to ask the user for permission for things.
+    initNotifications()
+    initSystemExtension()
+    initVPNConfiguration()
+  }
+
+  func initNotifications() {
     Task {
-      // Load user's decision whether to allow / disallow notifications
       self.decision = await self.sessionNotification.loadAuthorizationStatus()
+    }
+  }
 
-      // Load VPN configuration and system extension status
-      do {
-        try await self.bindToVPNConfigurationUpdates()
-        let vpnConfigurationStatus = self.status
-
+  func initSystemExtension() {
 #if os(macOS)
-        let systemExtensionStatus = try await self.checkedSystemExtensionStatus()
-
-        if systemExtensionStatus != .installed
-            || vpnConfigurationStatus == .invalid {
-
-          // Show the main Window if VPN permission needs to be granted
-          AppView.WindowDefinition.main.openWindow()
-        } else {
-          AppView.WindowDefinition.main.window()?.close()
-        }
+    Task {
+      do {
+        self.systemExtensionStatus = try await self.checkSystemExtensionStatus()
+      } catch {
+        Log.error(error)
+      }
+    }
 #endif
+  }
 
-        if vpnConfigurationStatus == .disconnected {
-
-          // Try to connect on start
-          try self.vpnConfigurationManager.start()
+  func initVPNConfiguration() {
+    Task {
+      do {
+        // Try to load existing configuration
+        if let neTunnelProviderManager = try await VPNConfigurationManager.loadFromPreferences() {
+          try await setupTunnelObservers(from: neTunnelProviderManager, autoStart: true)
+        } else {
+          status = .invalid
         }
       } catch {
         Log.error(error)
       }
     }
   }
 
-  public func internetResourceEnabled() -> Bool {
-    self.vpnConfigurationManager.internetResourceEnabled
-  }
+  func setupTunnelObservers(from manager: NETunnelProviderManager, autoStart: Bool = false) async throws {
+    guard let session = manager.connection as? NETunnelProviderSession
+    else { return }
```

Comment:
Instead of reaching into `NETunnelProviderManager`, this could be cleaner if it were a getter on `VPNConfigurationManager` that returns you the session if there is one.

To make it more encapsulated, you could even pass a (mutable) reference to the IPC client _into_ the constructor so the session could directly be initialised there, if there is one.

Reply:
👍